source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -0,0 +1,137 @@
+#include "rex_directx/directx_feature_level.h"
+#include "rex_directx/directx_feature_shader_model.h"
+#include "rex_directx/directx_util.h"
+#include "rex_directx/dxgi/adapter.h"
+#include "rex_directx/dxgi/adapter_manager.h"
+#include "rex_directx/dxgi/factory.h"
+#include "rex_directx/log.h"
+#include "rex_renderer_core/renderer.h"
+#include "rex_std/memory.h"
+#include "rex_std_extra/utility/enum_reflection.h"
+
+namespace rex
+{
+  namespace renderer
+  {
+    namespace internal
+    {
+      //-------------------------------------------------------------------------
+      const Gpu* highest_scoring_gpu(const rsl::vector<rsl::unique_ptr<Gpu>>& gpus)
+      {
+        auto it = std::max_element(gpus.cbegin(), gpus.cend(),
```

Comment:
should we assert that the gpus array length should be higher than 0?

Reply:
There is already an assert in the adaptor manager